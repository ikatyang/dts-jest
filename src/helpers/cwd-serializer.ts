export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
