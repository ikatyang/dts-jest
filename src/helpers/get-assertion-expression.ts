import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
