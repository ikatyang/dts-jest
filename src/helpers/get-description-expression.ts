import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
