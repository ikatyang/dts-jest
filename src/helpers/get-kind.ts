import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
