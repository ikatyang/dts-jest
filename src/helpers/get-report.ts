import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
