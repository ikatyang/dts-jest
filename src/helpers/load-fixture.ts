import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
