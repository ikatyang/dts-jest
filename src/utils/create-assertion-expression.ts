import { Trigger, TriggerFooterFlag, TriggerHeaderFlags } from '../definitions';

export interface CreateAssertionExpressionOptions {
  test_type: boolean;
  test_value: boolean;
  get_type_inference_or_diagnostic_expression: string;
  get_type_inference_or_throw_diagnostic_expression: string;
  get_type_report_expression: string;
  get_value_report_expression: string;
}

export const snapshot_assertion_message = '(type) should match snapshot';

export const create_assertion_expression = (
  trigger: Trigger,
  options: CreateAssertionExpressionOptions,
) => {
  const expressions: string[] = [];

  const { header, body, footer } = trigger;

  push_type_show_if_available();
  push_type_pass_if_available();
  push_type_fail_if_available();
  push_type_snap_if_available();
  push_type_not_any_if_available();
  push_value_show_if_available();
  push_value_equal_if_available();
  push_value_error_if_available();
  push_value_no_error_if_available();

  return expressions.join(';');

  function push_type_show_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':show']) {
      expressions.push(
        create_wrapper(
          '(type) should show report',
          `console.log(${options.get_type_report_expression})`,
        ),
      );
    }
  }

  function push_type_pass_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':pass']) {
      expressions.push(
        create_wrapper(
          '(type) should not throw error',
          `expect(function () { ${options.get_type_inference_or_throw_diagnostic_expression} }).not.toThrowError()`,
        ),
      );
    }
  }

  function push_type_fail_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':fail']) {
      expressions.push(
        create_wrapper(
          '(type) should throw error',
          `expect(function () { ${options.get_type_inference_or_throw_diagnostic_expression} }).toThrowError()`,
        ),
      );
    }
  }

  function push_type_snap_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':snap']) {
      expressions.push(
        create_wrapper(
          snapshot_assertion_message,
          `expect(${options.get_type_inference_or_diagnostic_expression}).toMatchSnapshot()`,
        ),
      );
    }
  }

  function push_type_not_any_if_available() {
    if (options.test_type && header.flags & TriggerHeaderFlags[':not-any']) {
      expressions.push(
        create_wrapper(
          '(type) should not be any',
          `expect(${options.get_type_inference_or_diagnostic_expression}).not.toBe("any")`,
        ),
      );
    }
  }

  function push_value_equal_if_available() {
    if (
      options.test_value &&
      footer !== undefined &&
      footer.flag === undefined
    ) {
      expressions.push(
        create_wrapper(
          `(value) should equal to ${footer.expected}`,
          `expect(${body.experssion}).toEqual(${footer.expected})`,
        ),
      );
    }
  }

  function push_value_error_if_available() {
    if (
      options.test_value &&
      footer !== undefined &&
      footer.flag === TriggerFooterFlag.Error
    ) {
      expressions.push(
        create_wrapper(
          '(value) should throw error',
          `expect(function () { ${body.experssion} }).toThrowError()`,
        ),
      );
    }
  }

  function push_value_no_error_if_available() {
    if (
      options.test_value &&
      footer !== undefined &&
      footer.flag === TriggerFooterFlag.NoError
    ) {
      expressions.push(
        create_wrapper(
          '(value) should not throw error',
          `expect(function () { ${body.experssion} }).not.toThrowError()`,
        ),
      );
    }
  }

  function push_value_show_if_available() {
    if (
      options.test_value &&
      footer !== undefined &&
      footer.flag === TriggerFooterFlag.Show
    ) {
      expressions.push(
        create_wrapper(
          '(value) should show report',
          `console.log(${options.get_value_report_expression})`,
        ),
      );
    }
  }

  function create_wrapper(title: string, expression: string) {
    const description = JSON.stringify(title);
    const {
      header: { method },
    } = trigger;
    return `${method}(${description}, function () { ${expression} })`;
  }
};
