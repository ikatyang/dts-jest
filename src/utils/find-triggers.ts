import * as _ts from 'typescript';
import { Trigger, TriggerFooter } from '../definitions';
import { find_trigger_bodies } from './find-trigger-bodies';
import { find_trigger_footers } from './find-trigger-footers';
import { find_trigger_headers } from './find-trigger-headers';
import { get_trigger_body_end_line } from './get-trigger-line';

export const find_triggers = (source_file: _ts.SourceFile, ts: typeof _ts) => {
  const triggers: Trigger[] = [];

  const headers = find_trigger_headers(source_file, ts);
  const bodies = find_trigger_bodies(source_file, headers, ts);
  const footers = find_trigger_footers(source_file, bodies, ts);

  for (let i = 0; i < headers.length; i++) {
    const header = headers[i];
    const body = bodies[i];
    let footer: TriggerFooter | undefined;

    const { line: body_end_line } = source_file.getLineAndCharacterOfPosition(
      body.end,
    );

    if (
      footers.length !== 0 &&
      body_end_line === get_trigger_body_end_line(footers[0].line)
    ) {
      footer = footers.shift();
    }

    triggers.push({
      header,
      body,
      footer,
    });
  }

  return triggers;
};
