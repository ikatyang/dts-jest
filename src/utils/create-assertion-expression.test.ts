import * as ts from 'typescript';
import { load_fixture_source_file } from '../helpers/load-fixture';
import { create_assertion_expression } from './create-assertion-expression';
import { find_triggers } from './find-triggers';

it('should return correctly', () => {
  const triggers = find_triggers(
    load_fixture_source_file('create-assertion-expression/example.ts', ts),
    ts,
  );
  expect(
    triggers.map(trigger => [
      trigger.body.expression,
      create_assertion_expression(trigger, {
        test_type: true,
        test_value: true,
        get_type_inference_or_diagnostic_expression:
          '<get_type_inference_or_diagnostic_expression>',
        get_type_inference_or_throw_diagnostic_expression:
          '<get_type_inference_or_throw_diagnostic_expression>',
        get_type_report_expression: '<get_type_report_expression>',
        get_value_report_expression: '<get_value_report_expression>',
      }).split(';'),
    ]),
  ).toMatchSnapshot();
});
