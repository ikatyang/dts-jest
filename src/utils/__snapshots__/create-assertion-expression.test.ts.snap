// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `
Array [
  Array [
    "Math.max(1, 2, 3)",
    Array [
      "test(\\"(type) should show report\\", function () { console.log(<get_type_report_expression>) })",
      "test(\\"(type) should not throw error\\", function () { expect(function () { <get_type_inference_or_throw_diagnostic_expression> }).not.toThrowError() })",
      "test(\\"(type) should match snapshot\\", function () { expect(<get_type_inference_or_diagnostic_expression>).toMatchSnapshot() })",
      "test(\\"(value) should equal to 3\\", function () { expect(Math.max(1, 2, 3)).toEqual(3) })",
    ],
  ],
  Array [
    "Math.abs(-1)",
    Array [
      "test(\\"(type) should throw error\\", function () { expect(function () { <get_type_inference_or_throw_diagnostic_expression> }).toThrowError() })",
      "test(\\"(value) should show report\\", function () { console.log(<get_value_report_expression>) })",
    ],
  ],
  Array [
    "Math.cos(Math.PI)",
    Array [
      "test(\\"(value) should not throw error\\", function () { expect(function () { Math.cos(Math.PI) }).not.toThrowError() })",
    ],
  ],
  Array [
    "Math.tan(Math.PI)",
    Array [
      "test(\\"(value) should throw error\\", function () { expect(function () { Math.tan(Math.PI) }).toThrowError() })",
    ],
  ],
  Array [
    "Math.sin(Math.PI)",
    Array [
      "test(\\"(type) should show report\\", function () { console.log(<get_type_report_expression>) })",
      "test(\\"(type) should not throw error\\", function () { expect(function () { <get_type_inference_or_throw_diagnostic_expression> }).not.toThrowError() })",
      "test(\\"(type) should match snapshot\\", function () { expect(<get_type_inference_or_diagnostic_expression>).toMatchSnapshot() })",
      "test(\\"(value) should show report\\", function () { console.log(<get_value_report_expression>) })",
    ],
  ],
  Array [
    "Object.assign({ a: 1 }, { b: 2 }, { c: 3 })",
    Array [
      "test(\\"(type) should match snapshot\\", function () { expect(<get_type_inference_or_diagnostic_expression>).toMatchSnapshot() })",
      "test(\\"(value) should equal to { a: 1, b: 2, c: 3, }\\", function () { expect(Object.assign({ a: 1 }, { b: 2 }, { c: 3 })).toEqual({ a: 1, b: 2, c: 3, }) })",
    ],
  ],
  Array [
    "Math.min(3, 2, 1)",
    Array [],
  ],
]
`;
